### ISSUE TITLE
[TFC] client side configs blocked

### ISSUE THREAD

**Issue by pizzahut2** [ValveSoftware/halflife#2108]
With today's update, class and map specific configs don't execute automatically any more with the following error message:

Could not execute privileged command exec mapdefault.cfg
Could not execute privileged command exec crossover2.cfg
Could not execute privileged command exec medic.cfg

AFAIK these configs should load if config.cfg contains these commands:

setinfo ec 1
setinfo em 1

or

setinfo exec_class on
setinfo exec_map on

Loading the class and map specific configs manually is still possible, but rather impractical.

**Comment by SamVanheer** [comment]
I recommend adding a new network message in game code that takes an enum indicating which file to execute:

```
int MsgFunc_SafeExec( const char *pszName, int iSize, void *pbuf )
{
	BEGIN_READ( pbuf, iSize );
	SafeExecFile fileId = ( SafeExecFile ) READ_BYTE();

        switch( fileId )
        {
        case SafeExecFile_Medic: gEngFuncs.pfnClientCmd( "medic.cfg" ); break;
        //Error handling omitted (check if level name is valid)
        case SafeExecFile_Map: gEngFuncs.pfnClientCmd( UTIL_VarArgs( exec mapdefault.cfg\nexec %s.cfg\n", gEngfuncs.pfnGetLevelName() ); break;
        //Additional cases omitted
        }

	return 1;
}
```

This solves the problem of arbitrary server-to-client command execution (if the `exec` command is also filtered, otherwise it will be an alternate attack vector).

If the exec command cannot be made secure, consider adding a new command `exec_filtered` that filters all commands parsed by it. This can then be used by server code in place of `exec`, perhaps implicitly by doing find and replace for commands checked by the filtering code:

```
if( doPrivilegeCheck && !strcmp( "exec", pszCommandName ) )
{
    pszCommandName = exec_filtered;
}
```

**Comment by SamVanheer** [comment]
This bug has been fixed by allowing the `exec` command for TFC only. Update your game installation to get the fix.

**Comment by pizzahut2** [comment]
Is the order of executing the map config files wrong? Should be

1. mapdefault.cfg
2. the cfg of the current map

With developer 1, I see this in console:

execing mapdefault.cfg
execing 2fort.cfg
my 2fort cfg loaded (echo in 2fort.cfg)
execing comms/map.cfg (This is executed by my 2fort.cfg.)
MAPDEFAULT CFG LOADED (echo in mapdefault.cfg)

So at first, it says correctly execing mapdefault.cfg and then 2fort.cfg, however the output shows that it actually executes them the other way round.

**Comment by DarthMan** [comment]
Exec must be allowed on any game, but only if the client has the file. For example, my servers are using a cfg max.cfg that's only executed on my server if players have it. It can be useful in many cases, as, for example, u want a key bound to something globally, but, on a specific server, you want it bound to something else. This is not slowhacking, since the player must have the file. Also, the client is the one who efits the file, so it's nkt like the server malciously does it. Hooe it will be re-enabled on all games.

**Comment by SamVanheer** [comment]
The problem with allowing exec is that servers can manipulate unsuspecting clients to download maps containing cfg files from a third party server that the server can then execute as a way to bypass protection. That's why i suggested filtering the commands parsed by the exec command.

**Comment by DarthMan** [comment]
Isn't cfg blocked from being downloaded?

**Comment by SamVanheer** [comment]
Yes, that's why i said third party server, as in "download the map here" type of thing. It's a form of social engineering.

**Comment by DarthMan** [comment]
Then Valve can just block downloading of these cfg files. This must be blocked by Valve as well https://gamebanana.com/projects/35384 . With this method, players can get VAC banned eithout them doing anything, just by connecting to a server that would inject a special client dll than will then be used on all kther servers, trigerring a vac ban on VAC secured servers, because, what if the dll the client gets has wallhack built ib?

**Comment by SamVanheer** [comment]
Valve can't block downloads that occur outside the game, which is what i'm referring to.

**Comment by i486** [comment]
Valve, Please hire this man! Make the CS 1.6 great again! :)

**Comment by pizzahut2** [comment]
Closing this issue, though the fix needs improvement as it caused a new issue as already mentioned above. https://github.com/ValveSoftware/halflife/issues/2112
